### DISCUSSION TITLE
How to remove a photo from a given album?

### DISCUSSION THREAD

**Discussion by shadowyman** [RhetTbull/osxphotos#1650] [Q&A] [answered]
I'm using the library to write my own duplicate finder which compiles metadata among dupes, picks ones with oldest date and crafts a generic metadata to be applied to best candidate. However, once I'm done, I'd like to remove all photos that were dupe of each other from the album. I know UUIDs and have photoInfo object in hand but can't seem to find anywhere how to remove assets from the album...

**Comment by oPromessa** [comment]
As per [PhotosAlbum](https://rhettbull.github.io/osxphotos/API_README.html#photosalbum) it seems not to be possible. 

> Assets cannot be removed from an album using this class due to limitations in the Photos AppleScript interface.

**Reply by oPromessa** [reply, marked as answer]
@shadowyman, Although with the helper tool from @RhetTbull  [PhotoScript](https://rhettbull.github.io/PhotoScript/#Limitations) he does mentions the `Album.remove` function/workaround. 

> Where possible, PhotoScript attempts to provide work-arounds to Photos’ limitations. For example, Photos does not provide a way to remove a photo from an album. PhotoScript does provide a Album.remove() method but in order to do this, it creates a new album with the same name as the original, copies all but the removed photos to the new album then deletes the original album. This simulates removing photos and produces the desired effect but is not the same thing as removing a photo from an album.

**Reply by shadowyman** [reply]
```python
import photoscript

# Activate the Photos app
photoslib = photoscript.PhotosLibrary()
photoslib.activate()

# Access the album
album_name = "Detected Duplicates3"
album = photoslib.album(album_name)

if album:
    # Find the photo by UUID
    photo_to_remove = None
    for photo in album.photos():
        if photo.uuid == "397F89B8-77F0-45C2-A03D-E0418":
            photo_to_remove = photo
            break

    if photo_to_remove:
        # Remove the photo by simulating the deletion
        album.remove([photo_to_remove])  # Pass the photo in a list
        print(f"Removed photo with UUID {photo_to_remove.uuid} from album '{album_name}'.")
    else:
        print(f"Photo with UUID '397F89B8-77F0-45C2-A03D-E0418' not found in album '{album_name}'.")
else:
    print(f"Album '{album_name}' not found.")

# Quit the Photos app
photoslib.quit()

```

Correct, I was just back to post this and found about your answer. I'm very close to finishing this project i'm working on: duplicate finder/metadata-exif compiler based on osxphotos and photoscript, which basically compares metadata among photos in a group of dupes and picks the oldest date (what i needed in my case) for date related exifdata: if say one photo of three dupes provides additional metadata, i also pick that up. this way i keep building an "ideal metada" for that image.

Then among dupes, I pick the image with highest size and resolution as the ideal candidate and using exiftool i'm writing ideal exifdata to that highest quality image. Right now it works about %95 of the total cases i'm working on. however my lib is 100k so i still can't tolerate 5%...

There's also that duplicates attr of photoInfo class in osxphotos, which seems to use basic criteria to find dupes. In my case this wasn't sensitive enough. Oh well, back to work...

**Reply by shadowyman** [reply]
Looks like this is more of a trick because album.photos() still return the photo removed by this method. Even though the photo doesn't appear under album in the app, it's still there when querying photos. I remember somewhere mentioned that same name albums can exist even though they are different. So this one is using that trick.

**Reply by RhetTbull** [reply]
If the goal is to eventually remove the photos themselves from the library, then what about just adding the duplicates you don't want to a new album ("Duplicates" for example) then when ready, delete them from the library which will remove them from the album.

**Reply by RhetTbull** [reply]
So if you don't mind using highly experimental "alpha" code, I have another project, [photokit](https://github.com/RhetTbull/photokit) which uses the native PhotoKit API from Apple (including a bunch of undocumented/private calls) to exert more control over Photos. This library *can* remove a photo from an album but there are some limitations; for example, you can't currently get the albums an asset is a member of from the album and so need to cycle through all the albums and checking to see if the asset is a member of the album. Sample code below to get you started.  ~It's not yet on PyPI so you'll need to install from GitHub.~ I had forgotten I had actually uploaded it...so you can `pip install photokit`. It is very alpha code at the moment.

This works on a single UUID but would be more efficient if you have a list of UUIDs to just batch them and make a single pass through the albums.

```python
"""Remove Photos asset with given UUID from all albums (experimental)"""

import sys
import time

import photokit


def remove_uuid_from_albums(uuid: str):
    """Remove asset with given UUID from all albums"""
    library = photokit.PhotoLibrary()
    assets = library.assets(uuids=[uuid])
    if not assets:
        raise ValueError(f"No asset found with UUID {uuid}")
    asset = assets[0]
    print(f"Removing asset {asset.original_filename} ({asset.uuid}) from all albums")
    for album in library.albums():
        # currently asset.albums() is not implemented so need to cycle through every album
        album_uuids = [a.uuid for a in album.assets()]
        if uuid in album_uuids:
            album.remove_assets([asset])
            print(f"Removed asset from album {album.title}")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <UUID>")
        sys.exit(1)
    while not photokit.PhotoLibrary.request_authorization():
        print("Waiting for authorization...")
        time.sleep(1)
    remove_uuid_from_albums(sys.argv[1])
```

**Comment by shadowyman** [comment]
Thanks much @RhetTbull  for the quick response. This could be really useful. However, I’m leaning towards the idea of moving photos around to different albums and deleting all photos in the album manually.

I’ve got few additional questions:

1. What’s the correct and safest way to delete an image from the library? I can’t see anything about it in osxphotos. 
2. If I got the path original of the photo from osxphotos, what will happen if I delete that file from photos library package? Will icloud update and delete thumbnail, everything about that photo?
3. What happens if i change metadata of the image with exifwriter and overwrite_file or overwrite_file_in_place flags? I’d like icloud to sync this metada change for the image across devices.

4. **Follow up to 3., what’s the safest method to changing metadata of the original photo? Should I use the original stored in photos library or export originals first then modify exported originals to take effect?**

Thanks a ton!

**Reply by shadowyman** [reply]
From my understanding, overwriting photo lib package originals with new metadata will cause issues. I think the correct way is exporting photos with photo.export() and write exif data while exporting. Then clean up dupes from albums (delete photos from album manually), and import these back in?

**Reply by RhetTbull** [reply]
> What’s the correct and safest way to delete an image from the library? I can’t see anything about it in osxphotos.

OSXPhotos cannot delete images. The best way is to move them to a "delete me" album then go in and delete them when you are ready.  The photokit library I referenced above does have a `delete_assets()` method but it will prompt the user and there is no way to bypass this. Either way you'll need to manually interact with Photos to delete images. This is enforced by Apple and I know of no way around it.

> If I got the path original of the photo from osxphotos, what will happen if I delete that file from photos library package? Will icloud update and delete thumbnail, everything about that photo?

Bad things would likely happen. I do not recommend this as it could lead to corruption of the library. At the very least, the asset would still appear in the library but Photos would complain when you tried to access it. It will not cause iCloud to delete the image.

> What h> appens if i change metadata of the image with exifwriter and overwrite_file or overwrite_file_in_place flags? I’d like icloud to sync this metada change for the image across devices.

`osxphotos push-exif` can do this for you but I would **not** do this with iCloud. This is only useful for libraries that do not sync to iCloud. Any changes made to original files in an iCloud library do not propagate to iCloud and are subject to being overwritten by Photos. 


> Follow up to 3., what’s the safest method to changing metadata of the original photo? Should I use the original stored in photos library or export originals first then modify exported originals to take effect?

If you want to change the metadata in the original file on disk (and iCloud), you need to export, modify the metadata and re-import. There is an open issue (#909) to automate this type of workflow but it's currently blocked waiting for me to complete the work on [photokit](https://github.com/RhetTbull/photokit).

**Reply by RhetTbull** [reply]
By the way, you can probably automate much of this with `osxphotos export --exiftool` to write the metadata once you clean up the duplicates then `osxphotos import --exportdb` to re-import and apply all the metadata (albums, keywords, etc.) from the exported images back in Photos.

**Reply by shadowyman** [reply]
Thanks Rhet, I'm planning to use photoInfo.export() function to export the best image (out of dupes) with

```python
exported_files = photo_info.export(
    "/path/to/export/directory",
    edited=True,
    exiftool=True
)
```

Then clean all images from the album manually. Once done, I'll manually import these best images. 

But I'm not sure if this actually supports passing in exif dictionary (with metadata). For that reason, it seems i need to use PhotoExporter class for this.

Thank you for all the help you've been!

**Reply by RhetTbull** [reply]
Yes, `PhotoInfo.export()` is intended for very simple use cases. [PhotoExporter](https://github.com/RhetTbull/osxphotos/blob/main/API_README.md#photoexporter) will give you much more control. It does not allow you to specify the exif dictionary though but it does allow you to control the final metadata written with exiftool using options such as `description_template`, `keyword_template`, etc. Alternatively, if you are crafting the EXIF dictionary yourself, you could export it then use [ExifTool](https://github.com/RhetTbull/osxphotos/blob/main/API_README.md#exiftool-1) to write the metadata after export.
